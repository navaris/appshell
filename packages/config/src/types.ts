import { JSONSchema4, JSONSchema6, JSONSchema7 } from 'json-schema';

export type Schema = JSONSchema4 | JSONSchema6 | JSONSchema7;

export type ConfigValidator = {
  validate: <T>(...config: T[]) => void;
};

export type ConfigMap = Record<string, string>;

/* appshell.config.yaml types */
export type AppshellConfigRemote<TMetadata = Record<string, unknown>> = {
  id: string;
  url: string;
  metadata: TMetadata;
};

export type AppshellConfig<TMetadata = Record<string, unknown>> = {
  name?: string;
  remotes?: Record<string, AppshellConfigRemote<TMetadata>>;
  module: ModuleFederationPluginOptions;
};

/** Appshell manifest types */
export type AppshellRemote<TMetadata = Record<string, unknown>> = {
  id: string;
  url: string;
  scope: string;
  module: string;
  shareScope?: string;
  metadata: TMetadata;
};

export type AppshellManifest<TMetadata = Record<string, unknown>> = {
  remotes: Record<string, AppshellRemote<TMetadata>>;
  modules: Record<string, ModuleFederationPluginOptions>;
};

/**
 * Advanced configuration for modules that should be exposed by this container.
 */
export interface ExposesConfig {
  /**
   * Request to a module that should be exposed by this container.
   */
  import: string | string[];

  /**
   * Custom chunk name for the exposed module.
   */
  name?: string;
}

export interface ExposesObject {
  [index: string]: string | ExposesConfig | string[];
}

/**
 * Set explicit comments for `commonjs`, `commonjs2`, `amd`, and `root`.
 */
export interface LibraryCustomUmdCommentObject {
  /**
   * Set comment for `amd` section in UMD.
   */
  amd?: string;

  /**
   * Set comment for `commonjs` (exports) section in UMD.
   */
  commonjs?: string;

  /**
   * Set comment for `commonjs2` (module.exports) section in UMD.
   */
  commonjs2?: string;

  /**
   * Set comment for `root` (global variable) section in UMD.
   */
  root?: string;
}

/**
 * Description object for all UMD variants of the library name.
 */
export interface LibraryCustomUmdObject {
  /**
   * Name of the exposed AMD library in the UMD.
   */
  amd?: string;

  /**
   * Name of the exposed commonjs export in the UMD.
   */
  commonjs?: string;

  /**
   * Name of the property exposed globally by a UMD library.
   */
  root?: string | string[];
}

/**
 * Options for library.
 */
export interface LibraryOptions {
  /**
   * Add a comment in the UMD wrapper.
   */
  auxiliaryComment?: string | LibraryCustomUmdCommentObject;

  /**
   * Specify which export should be exposed as library.
   */
  export?: string | string[];

  /**
   * The name of the library (some types allow unnamed libraries too).
   */
  name?: string | string[] | LibraryCustomUmdObject;

  /**
   * Type of library (types included by default are 'var', 'module', 'assign', 'assign-properties', 'this', 'window', 'self', 'global', 'commonjs', 'commonjs2', 'commonjs-module', 'commonjs-static', 'amd', 'amd-require', 'umd', 'umd2', 'jsonp', 'system', but others might be added by plugins).
   */
  type: string;

  /**
   * If `output.libraryTarget` is set to umd and `output.library` is set, setting this to true will name the AMD module.
   */
  umdNamedDefine?: boolean;
}

export interface SharedConfig {
  /**
   * Include the provided and fallback module directly instead behind an async request. This allows to use this shared module in initial load too. All possible shared modules need to be eager too.
   */
  eager?: boolean;

  /**
   * Provided module that should be provided to share scope. Also acts as fallback module if no shared module is found in share scope or version isn't valid. Defaults to the property name.
   */
  import?: string | false;

  /**
   * Package name to determine required version from description file. This is only needed when package name can't be automatically determined from request.
   */
  packageName?: string;

  /**
   * Version requirement from module in share scope.
   */
  requiredVersion?: string | false;

  /**
   * Module is looked up under this key from the share scope.
   */
  shareKey?: string;

  /**
   * Share scope name.
   */
  shareScope?: string;

  /**
   * Allow only a single version of the shared module in share scope (disabled by default).
   */
  singleton?: boolean;

  /**
   * Do not accept shared module if version is not valid (defaults to yes, if local fallback module is available and shared module is not a singleton, otherwise no, has no effect if there is no required version specified).
   */
  strictVersion?: boolean;

  /**
   * Version of the provided module. Will replace lower matching versions, but not higher.
   */
  version?: string | false;
}

export interface SharedObject {
  [index: string]: string | SharedConfig;
}

/**
 * Advanced configuration for container locations from which modules should be resolved and loaded at runtime.
 */
export interface RemotesConfig {
  /**
   * Container locations from which modules should be resolved and loaded at runtime.
   */
  external: string | string[];

  /**
   * The name of the share scope shared with this remote.
   */
  shareScope?: string;
}

/**
 * Container locations from which modules should be resolved and loaded at runtime. Property names are used as request scopes.
 */
export interface RemotesObject {
  [index: string]: string | RemotesConfig | string[];
}

export interface ModuleFederationPluginOptions {
  /**
   * Modules that should be exposed by this container. When provided, property name is used as public name, otherwise public name is automatically inferred from request.
   */
  exposes?: (string | ExposesObject)[] | ExposesObject;

  /**
   * The filename of the container as relative path inside the `output.path` directory.
   */
  filename?: string;

  /**
   * Options for library.
   */
  library?: LibraryOptions;

  /**
   * The name of the container.
   */
  name?: string;

  /**
   * The external type of the remote containers.
   */
  remoteType?:
    | 'import'
    | 'var'
    | 'module'
    | 'assign'
    | 'this'
    | 'window'
    | 'self'
    | 'global'
    | 'commonjs'
    | 'commonjs2'
    | 'commonjs-module'
    | 'commonjs-static'
    | 'amd'
    | 'amd-require'
    | 'umd'
    | 'umd2'
    | 'jsonp'
    | 'system'
    | 'promise'
    | 'script'
    | 'node-commonjs';

  /**
   * Container locations and request scopes from which modules should be resolved and loaded at runtime. When provided, property name is used as request scope, otherwise request scope is automatically inferred from container location.
   */
  remotes?: (string | RemotesObject)[] | RemotesObject;

  /**
   * The name of the runtime chunk. If set a runtime chunk with this name is created or an existing entrypoint is used as runtime.
   */
  runtime?: string | false;

  /**
   * Share scope name used for all shared modules (defaults to 'default').
   */
  shareScope?: string;

  /**
   * Modules that should be shared in the share scope. When provided, property names are used to match requested modules in this compilation.
   */
  shared?: (string | SharedObject)[] | SharedObject;
}
