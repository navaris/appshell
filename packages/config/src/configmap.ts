import { entries, set, values } from 'lodash';
import { AppshellConfig, ConfigMap } from './types';

const matchVariable = /(\${\w+})/;
const matchWord = /[^\w @]/g;

const findVariables = (val: string) =>
  val
    .split(matchVariable)
    .filter((w) => w.match(matchVariable))
    ?.map((w) => w.replace(matchWord, ''));

const findVariablePlaceholders = (
  obj: object | string | number | undefined,
  results: Record<string, string> = {},
): Record<string, string> =>
  values(obj).reduce((acc: Record<string, string>, val: object | string | number | undefined) => {
    if (typeof val === 'string') {
      const VARS = findVariables(val);
      return VARS.length > 0 ? VARS.reduce((a, v) => set<string>(a, v, process.env[v]), acc) : acc;
    }
    if (typeof val === 'object') {
      return findVariablePlaceholders(val, acc);
    }
    return acc;
  }, results);

const apply = <T extends object>(obj: T, configMap: ConfigMap): T => {
  entries(obj).forEach(([key, val]) => {
    if (typeof val === 'string') {
      const VARS = findVariables(val);
      if (VARS.length > 0) {
        VARS.forEach((v) => {
          const cur = obj[key];
          const value = configMap[v];
          if (!value) {
            // eslint-disable-next-line no-console
            console.log(`warning: value for ${v} is ${value}`);
          }
          set<string>(obj, key, cur.replace(`$\{${v}}`, configMap[v]));
        });
      }
    } else if (typeof val === 'object') {
      apply(val, configMap);
    }
  });

  return obj;
};

const create = (config: AppshellConfig) => findVariablePlaceholders(config);

export default {
  apply,
  create,
};
