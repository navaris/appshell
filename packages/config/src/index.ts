/**
 * @appshell/config package API
 */
export { default as configmap } from './configmap';
export { default as deregister } from './deregister';
export { default as generateEnv } from './generate.env';
export { default as generateGlobalConfig } from './generate.global-config';
export { default as generateManifest } from './generate.manifest';
export { default as register } from './register';
export type {
  AppshellConfig,
  AppshellConfigRemote,
  AppshellIndex,
  AppshellManifest,
  AppshellRemote,
  AppshellTemplate,
  Metadata,
  Schema,
} from './types';
export * as utils from './utils';
export * as validators from './validators';
