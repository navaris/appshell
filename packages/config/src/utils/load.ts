import fs from 'fs';
import yaml from 'yaml';

/**
 * Reads the YAML file
 * @param configPath to configuration file
 * @returns a yaml object
 */
const load = <TSchema>(configPath: string): TSchema => {
  if (!fs.existsSync(configPath)) {
    throw new Error(
      `Config file does not exist, using default config. __dirname: ${__dirname} Path: ${configPath}`,
    );
  }

  const configYAML = fs.readFileSync(configPath).toString();

  return yaml.parse(configYAML);
};

/**
 * Loads the YAML files specified by paths
 * @param paths Paths to configurations to process
 * @returns yaml objects
 */
export const loadAll = <TSchema>(...paths: string[]): TSchema[] => paths.map((path) => load(path));

export default load;
