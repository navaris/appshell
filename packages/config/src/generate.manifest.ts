/* eslint-disable no-param-reassign */
import fs from 'fs';
import path from 'path';
import configmap from './configmap';
import { toAppshellManifest } from './mappers/appshell.config';
import { AppshellTemplate } from './types';

/**
 * Generates an appshell manifest
 * @param templatePath path of the appshell manifest template to compile into an appshell manifest
 * @returns an appshell manifest
 */
export default async <TMetadata extends Record<string, unknown>>(templatePath: string) => {
  if (!fs.existsSync(templatePath)) {
    // eslint-disable-next-line no-console
    console.log(`No template file found '${templatePath}'`);

    return null;
  }

  const file = fs.readFileSync(path.resolve(templatePath), 'utf-8');
  const template = JSON.parse(file) as AppshellTemplate;
  const configMap = configmap.create(template);

  return toAppshellManifest<TMetadata>(template, configMap);
};
