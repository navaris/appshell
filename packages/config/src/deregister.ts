/* eslint-disable no-param-reassign */
import fs from 'fs';
import axios from './axios';
import { AppshellManifest, AppshellRegister } from './types';
import { isValidUrl } from './utils';
import * as validators from './validators';

const removeFromRegister = (doc: AppshellRegister, key: string) => {
  if (doc.index[key]) {
    delete doc.index[key];
  } else {
    // eslint-disable-next-line no-console
    console.warn(`index entry not found for '${key}'`);
  }

  if (doc.metadata && doc.metadata[key]) {
    delete doc.metadata[key];
  } else {
    // eslint-disable-next-line no-console
    console.warn(`metadata entry not found for '${key}'`);
  }

  return doc;
};

const removeFromManifest = (doc: AppshellManifest, key: string) => {
  const targetModule = key.split('/')[0];

  if (!doc.remotes[key]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for remotes[${key}]`);
  }

  if (!doc.modules[targetModule]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for modules[${key}]`);
  }

  if (!doc.environment[targetModule]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for environment[${key}]`);
  }

  delete doc.remotes[key];
  delete doc.modules[targetModule];
  delete doc.environment[targetModule];

  return doc;
};

const updateRegister = (registry: string, key: string) => {
  const file = `${registry}/appshell.register.json`;

  if (!fs.existsSync(file)) {
    // eslint-disable-next-line no-console
    console.warn(`registry file not found ${file}`);
    return;
  }

  const register: AppshellRegister = JSON.parse(fs.readFileSync(file, 'utf-8'));
  const doc = removeFromRegister(register, key);

  validators.appshell_register.validate(doc);

  fs.writeFileSync(file, JSON.stringify(doc));
};

const updateManifest = (registry: string, key: string) => {
  const file = `${registry}/appshell.manifest.json`;

  if (!fs.existsSync(file)) {
    // eslint-disable-next-line no-console
    console.warn(`registry file not found ${file}`);
    return;
  }

  const manifest: AppshellManifest = JSON.parse(fs.readFileSync(file, 'utf-8'));
  const doc = removeFromManifest(manifest, key);

  validators.merge_manifests.validate(doc);

  fs.writeFileSync(file, JSON.stringify(doc));
};

export default async (manifestKey: string, registry: string) => {
  // eslint-disable-next-line no-console
  console.log(`deregistering manifest from registry ${registry}`);

  if (isValidUrl(registry)) {
    await axios.delete(registry, { data: { manifestKey } });

    return;
  }

  if (!fs.existsSync(registry)) {
    // eslint-disable-next-line no-console
    console.warn(`registry location not found ${registry}`);
  }

  updateRegister(registry, manifestKey);
  updateManifest(registry, manifestKey);
};
