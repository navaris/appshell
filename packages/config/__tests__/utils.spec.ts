import chalk from 'chalk';
import fs from 'fs';
import path from 'path';
import { AppshellManifest } from '../src/types';
import { merge } from '../src/utils';
import copy from '../src/utils/copy';
import load from '../src/utils/load';
import validator from '../src/validators/AppshellManifestValidator';
import manifestA from './assets/appshell_manifests/test_a.manifest.json';
import manifestB from './assets/appshell_manifests/test_b.manifest.json';
import manifestC from './assets/appshell_manifests/test_c.manifest.json';
import manifestD from './assets/appshell_manifests/test_d.manifest.json';
import bizManifest from './assets/appshell_utils/bizmodule-biz.json';
import environmentCollisions from './assets/appshell_utils/environment-collisions.json';
import remoteCollisions from './assets/appshell_utils/remote-collisions.json';
import barManifest from './assets/appshell_utils/testmodule-bar.json';
import fooManifest from './assets/appshell_utils/testmodule-foo.json';

describe('utils', () => {
  const packageName = 'config';

  describe('load', () => {
    describe('consuming the configuration', () => {
      test('should read and parse the configuration file', () => {
        const file = path.resolve(`packages/${packageName}/__tests__/assets/appshell.config.yaml`);
        const config = load(file);

        expect(config).toMatchSnapshot();
      });

      test('should reject configuration file cannot be found', () => {
        const file = path.resolve(
          `packages/${packageName}/__tests__/assets/does_not_exist.config.yaml`,
        );

        expect(() => load(file)).toThrow(/Config file does not exist/);
      });
    });
  });

  describe('load', () => {
    describe('consuming the configuration', () => {
      test('should read and parse the configuration file', () => {
        const file = path.resolve(`packages/${packageName}/__tests__/assets/appshell.config.yaml`);
        const config = load(file);

        expect(config).toMatchSnapshot();
      });

      test('should reject configuration file cannot be found', () => {
        const file = path.resolve(
          `packages/${packageName}/__tests__/assets/does_not_exist.config.yaml`,
        );

        expect(() => load(file)).toThrow(/Config file does not exist/);
      });
    });
  });

  describe('copy', () => {
    const pattern = { from: '/path/to/file.json', to: '/new/path/to/file.json' };

    beforeAll(() => {
      jest.mock('fs');
    });

    afterAll(() => {
      jest.unmock('fs');
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    test('should not copy if source file does not exist', () => {
      const existsSyncSpy = jest.spyOn(fs, 'existsSync').mockReturnValueOnce(false);
      const mkdirSyncSpy = jest.spyOn(fs, 'mkdirSync');
      const copyFileSyncSpy = jest.spyOn(fs, 'copyFileSync');

      copy(pattern);

      expect(existsSyncSpy).toHaveBeenCalled();
      expect(mkdirSyncSpy).not.toHaveBeenCalled();
      expect(copyFileSyncSpy).not.toHaveBeenCalled();
    });

    test('should create the destination directory if it does not exist', () => {
      const toDir = path.dirname(pattern.to);
      const existsSyncSpy = jest
        .spyOn(fs, 'existsSync')
        .mockImplementation((file) => file === pattern.from);
      const mkdirSyncSpy = jest.spyOn(fs, 'mkdirSync').mockImplementationOnce(jest.fn());
      const copyFileSyncSpy = jest.spyOn(fs, 'copyFileSync').mockImplementation(jest.fn());

      copy(pattern);

      expect(existsSyncSpy).toHaveBeenCalledWith(pattern.from);
      expect(existsSyncSpy).toHaveBeenCalledWith(toDir);
      expect(mkdirSyncSpy).toHaveBeenCalledWith(toDir);
      expect(copyFileSyncSpy).toHaveBeenCalledWith(pattern.from, pattern.to);
    });

    test('should copy if source file exists', () => {
      jest.spyOn(fs, 'existsSync').mockReturnValue(true);
      const copyFileSyncSpy = jest.spyOn(fs, 'copyFileSync').mockImplementation(jest.fn());

      copy(pattern);

      expect(copyFileSyncSpy).toHaveBeenCalledWith(pattern.from, pattern.to);
    });
  });

  describe('merge', () => {
    let consoleSpy: jest.SpyInstance;
    beforeEach(() => {
      consoleSpy = jest.spyOn(console, 'log');
    });

    afterEach(() => {
      consoleSpy.mockRestore();
    });

    test('should merge multiple valid configurations', () => {
      const config = merge<AppshellManifest>(validator, fooManifest, barManifest, bizManifest);
      expect(config).toMatchSnapshot();
    });

    test('should merge manifests from right to left with merge validator', () => {
      const config = merge<AppshellManifest>(validator, manifestA, manifestB, manifestC, manifestD);

      expect(config).toMatchSnapshot();
    });

    test('should reject when no validator is found for schema', () => {
      expect(() =>
        merge<AppshellManifest>(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          undefined as any,
          fooManifest,
          barManifest,
          bizManifest,
        ),
      ).toThrow(/No validator provided/);
    });

    test('should validate add documents individually', () => {
      const validateSpy = jest.spyOn(validator, 'validate');
      merge<AppshellManifest>(validator, fooManifest, barManifest, bizManifest);

      expect(validateSpy).toHaveBeenCalledWith(fooManifest);
      expect(validateSpy).toHaveBeenCalledWith(barManifest);
      expect(validateSpy).toHaveBeenCalledWith(bizManifest);
    });

    test('should validate merged document', () => {
      const validateSpy = jest.spyOn(validator, 'validate');
      const config = merge<AppshellManifest>(validator, fooManifest, barManifest, bizManifest);

      expect(validateSpy).toHaveBeenLastCalledWith(config);
    });

    describe('appshell configs', () => {
      test('should merge multiple valid configurations', () => {
        const config = merge<AppshellManifest>(validator, fooManifest, barManifest, bizManifest);

        expect(config).toMatchSnapshot();
      });

      test('should reject configurations with remotes collisions', () => {
        merge<AppshellManifest>(validator, fooManifest, barManifest, remoteCollisions);

        expect(consoleSpy).toHaveBeenCalledWith(chalk.yellow('Multiple remotes with the same key'));
      });

      test('should reject configurations with environment collisions', () => {
        merge<AppshellManifest>(validator, fooManifest, barManifest, environmentCollisions);

        expect(consoleSpy).toHaveBeenCalledWith(
          chalk.yellow('Multiple environments with the same key'),
        );
      });
    });
  });
});
