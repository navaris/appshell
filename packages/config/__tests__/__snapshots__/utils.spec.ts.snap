// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils load consuming the configuration should read and parse the configuration file 1`] = `
{
  "remotes": {
    "BizModule/Biz": {
      "filename": "remoteEntry.js",
      "metadata": {
        "displayGroup": "auxiliary",
        "displayName": "Biz App",
        "icon": "ViewList",
        "order": 30,
        "route": "/biz",
      },
      "url": "http://localhost:4040",
    },
    "TestModule/Bar": {
      "filename": "remoteEntry.js",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Bar App",
        "icon": "ViewList",
        "order": 20,
        "route": "/bar",
      },
      "url": "\${APPS_TEST_URL}",
    },
    "TestModule/Foo": {
      "filename": "remoteEntry.js",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "url": "\${APPS_TEST_URL}",
    },
  },
}
`;

exports[`utils load consuming the configuration should read and parse the configuration file 2`] = `
{
  "remotes": {
    "BizModule/Biz": {
      "filename": "remoteEntry.js",
      "metadata": {
        "displayGroup": "auxiliary",
        "displayName": "Biz App",
        "icon": "ViewList",
        "order": 30,
        "route": "/biz",
      },
      "url": "http://localhost:4040",
    },
    "TestModule/Bar": {
      "filename": "remoteEntry.js",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Bar App",
        "icon": "ViewList",
        "order": 20,
        "route": "/bar",
      },
      "url": "\${APPS_TEST_URL}",
    },
    "TestModule/Foo": {
      "filename": "remoteEntry.js",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "url": "\${APPS_TEST_URL}",
    },
  },
}
`;

exports[`utils merge appshell configs should merge multiple valid configurations 1`] = `
{
  "environment": {
    "biz_module": {
      "RUNTIME_ENV": "Biz env",
      "RUNTIME_ENV_VERSION": "Biz version",
    },
    "test_module_bar": {
      "RUNTIME_ENV": "Test env",
      "RUNTIME_ENV_VERSION": "Test version",
    },
    "test_module_foo": {
      "RUNTIME_ENV": "Foo env",
      "RUNTIME_ENV_VERSION": "Foo version",
    },
  },
  "modules": {
    "BizModule": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "BizModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
    "TestModuleBar": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModuleBar",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
    "TestModuleFoo": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModuleBar",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
  },
  "remotes": {
    "BizModule/Biz": {
      "id": "BizModule/Biz",
      "manifestUrl": "http://localhost:4040/appshell.manifest.json",
      "metadata": {
        "displayGroup": "auxiliary",
        "displayName": "Biz App",
        "icon": "ViewList",
        "order": 30,
        "route": "/biz",
      },
      "module": "./Biz",
      "remoteEntryUrl": "http://localhost:4040/remoteEntry.js",
      "scope": "BizModule",
    },
    "TestModuleBar/Bar": {
      "id": "TestModule/Bar",
      "manifestUrl": "http://localhost:4040/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Bar App",
        "icon": "ViewList",
        "order": 20,
        "route": "/bar",
      },
      "module": "./Bar",
      "remoteEntryUrl": "http://localhost:4040/remoteEntry.js",
      "scope": "TestModule",
    },
    "TestModuleFoo/Foo": {
      "id": "TestModule/Foo",
      "manifestUrl": "http://localhost:4040/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "module": "./Foo",
      "remoteEntryUrl": "http://localhost:4040/remoteEntry.js",
      "scope": "TestModule",
    },
  },
}
`;

exports[`utils merge should merge manifests from right to left with merge validator 1`] = `
{
  "environment": {
    "biz_module": {
      "RUNTIME_ENV": "Biz env",
      "RUNTIME_ENV_VERSION": "Biz version",
    },
    "test_module_a": {
      "RUNTIME_ENV": "A env",
      "RUNTIME_ENV_VERSION": "A version",
    },
    "test_module_b": {
      "RUNTIME_ENV": "B env",
      "RUNTIME_ENV_VERSION": "B version",
    },
    "test_module_c": {
      "RUNTIME_ENV": "C env",
      "RUNTIME_ENV_VERSION": "C version",
    },
  },
  "modules": {
    "TestModule": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
    "TestModuleA": {
      "exposes": {
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
    "TestModuleC": {
      "exposes": {
        "./Bang": "./src/Entry3",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
  },
  "overrides": {
    "environment": {
      "test_module_a": {
        "RUNTIME_ENV": "A env override",
      },
    },
  },
  "remotes": {
    "BizModule/Biz": {
      "id": "BizModule/Biz",
      "manifestUrl": "http://localhost:4040/appshell.manifest.json",
      "metadata": {
        "displayGroup": "auxiliary",
        "displayName": "Biz App",
        "icon": "ViewList",
        "order": 30,
        "route": "/biz",
      },
      "module": "./Biz",
      "remoteEntryUrl": "http://localhost:4040/remoteEntry.js",
      "scope": "BizModule",
    },
    "TestModuleA/Foo": {
      "id": "TestModule/Foo",
      "manifestUrl": "http://localhost:8080/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "remoteEntryUrl": "http://localhost:8080/remoteEntry.js",
    },
    "TestModuleB/Bar": {
      "id": "TestModule/Bar",
      "manifestUrl": "http://localhost:8080/bar/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "remoteEntryUrl": "http://localhost:8080/bar/remoteEntry.js",
    },
    "TestModuleB/Foo": {
      "id": "TestModule/Foo",
      "manifestUrl": "http://localhost:8080/foo/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 20,
        "route": "/foo",
      },
      "remoteEntryUrl": "http://localhost:8080/foo/remoteEntry.js",
    },
    "TestModuleC/Foo": {
      "id": "TestModule/Foo",
      "manifestUrl": "http://localhost:8080/foo_c/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 30,
        "route": "/foo_c",
      },
      "remoteEntryUrl": "http://localhost:8080/foo_c/remoteEntry.js",
    },
  },
}
`;

exports[`utils merge should merge multiple valid configurations 1`] = `
{
  "environment": {
    "biz_module": {
      "RUNTIME_ENV": "Biz env",
      "RUNTIME_ENV_VERSION": "Biz version",
    },
    "test_module_bar": {
      "RUNTIME_ENV": "Test env",
      "RUNTIME_ENV_VERSION": "Test version",
    },
    "test_module_foo": {
      "RUNTIME_ENV": "Foo env",
      "RUNTIME_ENV_VERSION": "Foo version",
    },
  },
  "modules": {
    "BizModule": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "BizModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
    "TestModuleBar": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModuleBar",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
    "TestModuleFoo": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModuleBar",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
  },
  "remotes": {
    "BizModule/Biz": {
      "id": "BizModule/Biz",
      "manifestUrl": "http://localhost:4040/appshell.manifest.json",
      "metadata": {
        "displayGroup": "auxiliary",
        "displayName": "Biz App",
        "icon": "ViewList",
        "order": 30,
        "route": "/biz",
      },
      "module": "./Biz",
      "remoteEntryUrl": "http://localhost:4040/remoteEntry.js",
      "scope": "BizModule",
    },
    "TestModuleBar/Bar": {
      "id": "TestModule/Bar",
      "manifestUrl": "http://localhost:4040/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Bar App",
        "icon": "ViewList",
        "order": 20,
        "route": "/bar",
      },
      "module": "./Bar",
      "remoteEntryUrl": "http://localhost:4040/remoteEntry.js",
      "scope": "TestModule",
    },
    "TestModuleFoo/Foo": {
      "id": "TestModule/Foo",
      "manifestUrl": "http://localhost:4040/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "module": "./Foo",
      "remoteEntryUrl": "http://localhost:4040/remoteEntry.js",
      "scope": "TestModule",
    },
  },
}
`;
